snapcraft/plugins/jhbuild.py

```diff
@@ -0,0 +1,383 @@
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""The JHBuild plugin is used for GNOME-based parts.
+
+JHBuild is a tool used to build components of the GNOME ecosystem from version
+control. See the JHBuild manual for more details:
+https://developer.gnome.org/jhbuild/stable/.
+
+The plugin can be customized with the following keys:
+
+  - modules: (required) a list of modules to include in the part, modules that
+             should be skipped should be prefixed with '-', e.g. '-WebKit'
+  - module-set: (default: gnome-world) the module set JHBuild should use
+  - module-set-dir: the directory where the module set is located
+               Either leave unset to use the default moduleset definitions, or
+             set to `.` to reference your project folder.
+  - jhbuild-archive: the source tarball directory on the build host
+  - jhbuild-mirror: the DVCS repository directory on the build host
+
+Advice:
+
+  - You only need to specify a single module in most cases. JHBuild knows the
+    module dependency graph.
+
+  - The desktop-gtk3 helper is incompatible with GTK built by jhbuild. A custom
+    helper is currently required.
+    See github.com/diddledan/corebird for example.
+
+  - Building WebKit requires a lot of time and computing power. If WebKit is a
+    dependency of your JHBuild module, it can be skipped by adding '-WebKit' to
+    the list of modules, and adding 'libwebkit2gtk-4.0-dev' to the
+    build-packages and stage-packages.
+
+  - Specify directories on the build host for jhbuild-archive and
+    jhbuild-mirror to prevent repeated downloading of the JHBuild module
+    sources. It's best to reserve common directories on your local machine that
+    can be reused by all snaps you might want to build.
+"""
+
+import glob
+import logging
+import os
+import subprocess
+import snapcraft
+from snapcraft.internal import common
+
+HOME = os.path.expanduser('~') if os.geteuid() else os.path.join('/', 'root')
+
+BUILD_PACKAGES = {
+    'apt-file',
+    'autoconf',
+    'automake',
+    'autopoint',
+    'autotools-dev',
+    'bison',
+    'build-essential',
+    'ca-certificates',
+    'cvs',
+    'docbook',
+    'docbook-xml',
+    'docbook-xsl',
+    'flex',
+    'gettext',
+    'git',
+    'intltool',
+    'iso-codes',
+    'libtext-csv-perl',
+    'libtool',
+    'libxml-parser-perl',
+    'make',
+    'ninja-build',
+    'pkg-config',
+    'pkgconf',
+    'python-dev',
+    'python3-dev',
+    'subversion',
+    'symlinks',
+    'yelp-tools',
+    'zlib1g-dev',
+}
+
+
+LOG = logging.getLogger(__name__)
+
+
+def _get_jhbuild_user():
```

Comment:
a docstring in this function would be useful, because it's not clear for me why it's needed.